libs/wingsdk/src/core/plugin-manager.ts

```diff
@@ -0,0 +1,177 @@
+
+    for (const plugin of plugins) {
+      this.add(plugin);
+    }
+  }
+
+  /**
+   * Add plugins to the plugin manager based on the plugin type.
+   *
+   * @param plugin the plugin to add
+   */
+  public add(plugin: string) {
+    // maybe we support other plugin types in the future (e.g. npm modules)
+    if (!plugin.endsWith(".js")) {
+      throw new Error(
+        `Currently only javascript files are supported as plugins. Got: ${plugin}`
+      );
+    }
+    this.addPluginFromFilePath(plugin);
+  }
+
+  /**
+   * Loads hooks from a plugin file and adds them to the list of hooks
+   * that will later be run in the various plugin phases.
+   *
+   * @param filePath absolute file path to the plugin
+   */
+  private addPluginFromFilePath(filePath: string) {
+    const pluginHooks: ICompilationHook = {
+      pluginName: filePath, // can be overwritten by plugin (ex: exports.pluginName = "my-plugin")
+    };
+
+    const pluginDir = resolve(filePath).split("/").slice(0, -1).join("/");
+
+    const context = vm.createContext({
+      require,
+      console,
+      exports: pluginHooks,
+      process,
+      __dirname: pluginDir,
+    });
+
+    const pluginCode = readFileSync(resolve(filePath), "utf8");
+    const script = new vm.Script(pluginCode);
+    script.runInContext(context);
+
+    this.hooks.push(pluginHooks);
+  }
+
+  private throwHookInvocationError(
+    stage: string,
+    pluginName: string,
+    err: any
+  ) {
+    throw new Error(
+      `Plugin: "${pluginName}" failed, during stage: "${stage}". cause: ${err.message}`
+    );
+  }
+
+  /**
+   * Call all preSynth hooks
+   *
+   * This phase is mutable, allowing plugins to modify the app construct
+   * before any synthesis takes place.
+   *
+   * @param app app construct to be passed to the preSynth hooks
+   */
+  public preSynth(app: IConstruct) {
+    for (const hook of this.hooks) {
+      if (hook.preSynth) {
+        try {
+          hook.preSynth(app);
+        } catch (err) {
+          this.throwHookInvocationError(
+            PluginPhases.PRE_SYNTH,
+            hook.pluginName,
+            err
+          );
+        }
+      }
+    }
+  }
+
+  /**
+   * Call all postSynth hooks
+   *
+   * This phase is mutable, allowing plugins to modify the Terraform config
+   * after synthesis has occurred then overwrite the config file on disk
+   * with the modified one.
+   *
+   * @param config Terraform config to be passed to the postSyth hooks
+   * @param synthesizedStackPath path to the synthesized stack json file
+   */
+  public postSynth(config: any, synthesizedStackPath: string): any {
+    for (const hook of this.hooks) {
+      if (hook.postSynth) {
+        try {
+          config = hook.postSynth(config) ?? config; // ignores undefined return values
+        } catch (err) {
+          this.throwHookInvocationError(
+            PluginPhases.POST_SYNTH,
+            hook.pluginName,
+            err
+          );
+        }
+      }
+    }
```

Comment:
Can we reuse this pattern instead of copy&pasting for every hook?

Something like:

```js
return this.callAllHooks(
  PluginPhases.POST_SYNTH, 
  hook => hook.postSynth(config) ?? config
);
```